src/Maestro/Maestro.Web/Api/v2019_01_16/Controllers/BuildTimeController.cs

```diff
@@ -0,0 +1,128 @@
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+
+
+namespace Maestro.Web.Api.v2019_01_16.Controllers
+{
+    /// <summary>
+    ///   Exposes methods to Read <see cref="BuildTime"/>s
+    /// </summary>
+    [Route("buildtime")]
+    [ApiVersion("2019-01-16")]
+    public class BuildTimeController : ControllerBase
+    {
+        protected readonly BuildAssetRegistryContext _context;
+        private readonly IKustoClientProvider _kustoClientProvider;
+
+        public BuildTimeController(
+            BuildAssetRegistryContext context,
+            IKustoClientProvider kustoClientProvider)
+        {
+            _context = context;
+            _kustoClientProvider = kustoClientProvider;
+        }
+
+        /// <summary>
+        /// Gets the average official build time and average pr build time for a given default channel
+        /// This is captured for generating the longest build times for the dependency flow graph
+        /// </summary>
+        /// <param name="id">Default channel id</param>
+        [HttpGet("{id}")]
+        [SwaggerApiResponse(HttpStatusCode.OK, Type = typeof(BuildTime), Description = "Gets the average official build time and average pr build time for a given default channel")]
+        [ValidateModelState]
+        public virtual async Task<IActionResult> GetBuildTimes([Required]int id)
+        {
+            Data.Models.DefaultChannel defaultChannel = await _context.DefaultChannels.FindAsync(id);
+
+            if (defaultChannel == null)
+            {
+                return NotFound();
+            }
+
+            var properties = new ClientRequestProperties();
+
+            // There is a mismatch between how we store the names of the repositories in BAR
+            // and how we store them in Kusto. Normalize to the last part of the url.
+            properties.SetParameter("_Repository", defaultChannel.Repository.Split("/").Last());
+            properties.SetParameter("_SourceBranch", defaultChannel.Branch);
+
+            string queryParameters = String.Join(",", properties.Parameters.Select(p => $"{p.Key}:{KustoDataTypes.String}"));
+
+            // We only care about builds that complete successfully or partially successfully 
+            // from the given repository. We summarize duration of the builds over the last 7 days.
+            // There are multiple different definitions that run in parallel, so we summarize
+            // on the definition id and ultimately choose the definition that took the longest.
+            string commonQueryText = @"| where Repository endswith _Repository
+                | where Result != 'failed' and Result != 'canceled' 
+                | where FinishTime > ago(7d) 
+                | extend duration = FinishTime - StartTime 
+                | summarize avg(duration) by DefinitionId
+                | summarize max(avg_duration)";
+
+            // We only want the pull request time from the public ci. We don't exclude on branch,
```

Comment:
Would this end up mixing up PR builds against other branches? e.g. if 3.0 and 3.1 have really slow PRs for aspnetcore, but 5 has a fast PR, would we end up taking the average of all those together?

Reply:
I wonder whether we need to alter our kusto telemetry to record the target branch of a PR.

Reply:
That is correct. I don't know how to tell what branch a PR applies to, so the PR time will be a little off. Do you have any suggestions on how to tell what branch a PR would apply to?

Reply:
I don't know if this is a problem on our side. I think AzDO only knows the merge branch name, and doesn't actually know where it's a pr to, based on the AzureDataTypes we have in AzureDevOpsTimeline.

Reply:
The AzDO PR info certainly knows the target branch, we just don't record it.

Reply:
In order to have accurate data, we probably need to add that column to the timeline data for PRs.

Reply:
It's not in the timeline result, so we'll need to make another API call for every build to determine it's branch.  I don't see any API that actually has that information available.  The "Timelines" API doesn't have any branch information, and the "Builds" API only has the SourceBranch, not the TargetBranch.

Normally you'd look at the PR, but there is not AzDO PR to fine here, so we might have to cross reference into GitHub to get this information?

That... might be fine... but this doesn't seem like this particular feature justifies that cost. My understanding is that PRs are generally overwhelmingly only to a single branch, so the data isn't going to be far off.

Reply:
You certainly could here.  The merge branches contain the PR number, so looking at the reason + PR number from source branch + source branch location (github) could enable you to find the PR involved and get the target branch.

I don't know if I buy that the PR times will generally be close once we move into the .net 5 timeframe.  Some repos are consolidating here (e.g. cli master in 3.x is not the same repo as it is in 5)

Reply:
I'm not sure the AzDO build knows anything at all about the target branch. It directly checks out the merge branch from github/azdo, as far as I can tell.

If we want this info, I think we'd have to figure it out in code in the AzureDevOpsTimeline app, and go query the github api for all public PRs (the only ones we care about) we are ingesting into Kusto, find the base section of the response, and then ref. Something like that. I feel like that should be done outside the scope of this PR, but I understand that we would want that functionality before this change gets approved.

Reply:
> I'm not sure the AzDO build knows anything at all about the target branch. It directly checks out the merge branch from github/azdo, as far as I can tell.

Yeah the timeline data doesn't, but you can take the repo + the source branch (which contains the PR number) and use that to look up a PR.

Reply:
I also remembered while talking with Epsi and Megan than github throttles super hard when you try to programmatically ping their api too frequently (at least it has in the past for me).

Reply:
I think when a build comes from GitHub, the "triggerInfo" return from the Build has a "pr.number" and "pr.providerId" == "github", so we avoid parsing the branch name (since that's not really documented anywhere, so it seems like that might be fragile).  So that at least gives us the PR number.

Though the "parameters" of the Build also appears to have "system.pullRequest.targetBranch" in it... which points to "master"... but it's a string encoded JSON blob, so I'm not 100% sure that's safe to depend on either. :-)

Reply:
> I also remembered while talking with Epsi and Megan than github throttles super hard when you try to programmatically ping their api too frequently (at least it has in the past for me).

We have a rate limit of about 12.5k requests/hour for this app. it should be plenty as long as we don't request the same data over and over.

Reply:
> Though the "parameters" of the Build also appears to have "system.pullRequest.targetBranch" in it... which points to "master"... but it's a string encoded JSON blob, so I'm not 100% sure that's safe to depend on either. :-)

It might be safe to depend on that and throw if it is not in the expected form.  In the general vein of public API design: "if you do it, someone depends on it". I think we'd unlikely to be broken. If we were, having appropriate error handling in place would ensure we know about it.

Reply:
Based on this discussion, I'm thinking I can do something like this when we ingest the data to kusto:

```
if (build.Reason == "pullRequest")
                {
                    var buildParameters = JsonConvert.DeserializeObject<Dictionary<string,string>>(build.Parameters);
                    targetBranch = buildParameters.ContainsKey("system.pullRequest.targetBranch") ? buildParameters["system.pullRequest.targetBranch"] : "";
                }
```

And add the TargetBranch column to the TimelineBuilds table. Sound reasonable?

Reply:
I'd probably just leave it a "JObject", in case something weird ends up in there, like a nested object, or a number, we wouldn't want it to fail.

So you could do `targetBranch = JObject.Parse(build.Parameters)["system.pullRequest.targetBranch"]`. JSON.NET returns null if you ask for a property that doesn't exist.

I might put it in a try/catch too and _log.LogError the exception, in case the build.Parameters is malformed for some reason (we wouldn't want to fail the whole batch and get stuck in a loop because of a single bad value here)

Reply:
done.